### ISSUE TITLE
Override problem

### ISSUE THREAD

**Issue by rfranco** [mohiva/play-silhouette#336]
I tried to override the DummyAuthenticatorService but it give a weird error.

```scala
class OverrideAuthService extends DummyAuthenticatorService {
  override def retrieve(implicit request: RequestHeader) = {
    Future.successful(Some(DummyAuthenticator(LoginInfo("abc", "123"))))
  }
}
```

```bash
[error]  found   : Some[com.mohiva.play.silhouette.impl.authenticators.DummyAuthenticator]
[error]  required: None.type
[error]     Future.successful(Some(DummyAuthenticator(LoginInfo("abc", "123"))))
```

It's the current implemantation
```scala
class DummyAuthenticatorService extends AuthenticatorService[DummyAuthenticator] {
  def retrieve(implicit request: RequestHeader) = Future.successful(None)
}
```

It only works when add the return type

```scala 
class DummyAuthenticatorService extends AuthenticatorService[DummyAuthenticator] {
  override def retrieve(implicit request: RequestHeader):Future[Option[DummyAuthenticator]] = Future.successful(None)
}
```

@akkie Should we add the return type for all methods?

**Comment by akkie** [comment]
We do not need define the return type for every method. Only for methods which have defined a container type like `Option` or `Either`. And only if the compiler cannot infer the correct type. If I understand it correctly then Scala allows an overridden method to have a covariant return type. So in the base class the compiler infers `None` as return type because it's a subclass of `Option`. But now you try to override the return type `None` with `Some`, which ends in a compiler error, because there is no covariance between the both types.

**Comment by rfranco** [comment]
Ok, i'm going review and make the changes then i submit a pull request.

**Comment by joaoraf** [comment]
IMHO,
I think that public and protected methods and fields should be explicitly type as a form of contract and documentation. Private ones don't have to be.

**Comment by akkie** [comment]
Fixed with #354
